### ISSUE TITLE
OpenGL: Texture rendering error

### ISSUE THREAD

**Issue by 256colour** [adventuregamestudio/ags#574]
Creating a new issue to address the texture scaling problem noted [here](https://github.com/adventuregamestudio/ags/issues/510#issuecomment-450741174), [here](https://github.com/adventuregamestudio/ags/issues/510#issuecomment-423166151), [here](https://github.com/adventuregamestudio/ags/issues/510#issuecomment-423465358), and [here](https://github.com/adventuregamestudio/ags/issues/510#issuecomment-423502288).

Behaviour:

- When `render_at_screenres = 1` in acsetup.cfg, a thin line (sometimes black, or other colours) is visible at the bounding box of character sprites.
- setting `render_at_screenres = 0` eliminates the problem
- reported to occur on both Windows and Linux OpenGL builds.

Research:

- @ivan-mogilko [notes that texture scaling may be the underlying problem](https://github.com/adventuregamestudio/ags/issues/510#issuecomment-423684066).
- @toojays [replicated the problem here](https://github.com/adventuregamestudio/ags/issues/510#issuecomment-451105599) in Gemini Rue.
- Setting ```glDisable(GL_DITHER);``` did not help according to @toojays, based on a recommendation from [this stackexchange thread](https://gamedev.stackexchange.com/questions/20053/what-is-causing-these-visual-artifacts-on-my-opengl-sprites).

**Comment by ivan-mogilko** [comment]
Alright, I spent an hour trying to reproduce this in different resolutions and scaling sizes to no avail, and then changed graphics filter to "Linear" and immediately got this error.

Can you confirm (or deny) that this happens only with linear filter to you?

**Comment by ivan-mogilko** [comment]
This may be a bit hacky and I would look a bit more into this, but for now I was able to fix this by adding following code:
<pre>
  if (tile->width < tileWidth)
  {
      for (int y = 0; y < tileHeight; y++)
      {
          unsigned int* memPtrLong = (unsigned int*)(memPtr + y * pitch + tile->width * sizeof(int));
          unsigned int* memPtrLong_previous = (unsigned int*)(memPtr + y * pitch + (tile->width - 1) * sizeof(int));
          *memPtrLong = *memPtrLong_previous & 0x00FFFFFF;
      }
  }
</pre>

At the following location: https://github.com/adventuregamestudio/ags/blob/ags3/Engine/gfx/ali3dogl.cpp#L1546

Also, it's interesting that for iOS and Android there is this redefinition:
<pre>#define GL_CLAMP GL_CLAMP_TO_EDGE</pre>

I tried doing same for Windows but it did not have any observable effect.

**Comment by 256colour** [comment]
Interesting, I see this code already in the Mac OpenGL implementation "osx-nick" branch in Janet's fork of AGS:

```
  for (int y = 0; y < tileHeight; y++)
  {
    // Mimic the behaviour of GL_CLAMP_EDGE for the bottom line
    if (y == tile->height)
    {
      unsigned int* memPtrLong = (unsigned int*)memPtr;
      unsigned int* memPtrLong_previous = (unsigned int*)(memPtr - tileWidth * 4);

      for (int x = 0; x < tileWidth; x++)
        memPtrLong[x] = memPtrLong_previous[x] & 0x00FFFFFF;

      continue;
    }
```

**Comment by 256colour** [comment]
I'm working on the Mac OpenGL build of the game found in the ["osx-nick" branch of Janet's AGS fork](https://github.com/JanetGilbert/ags/tree/osx-nick). I'm trying to integrate the patch, but I could use some help with modifying the method - it seems to have a slightly different implementation than the one your example uses:

`void OGLGraphicsDriver::UpdateTextureRegion(TextureTile *tile, Bitmap *bitmap, OGLBitmap *target, bool hasAlpha)
{
  int textureHeight = tile->height;
  int textureWidth = tile->width;

  AdjustSizeToNearestSupportedByCard(&textureWidth, &textureHeight);

  int tileWidth = (textureWidth > tile->width) ? tile->width + 1 : tile->width;
  int tileHeight = (textureHeight > tile->height) ? tile->height + 1 : tile->height;

  bool usingLinearFiltering = _filter->UseLinearFiltering();
  bool lastPixelWasTransparent = false;
  char *origPtr = (char*)malloc(4 * tileWidth * tileHeight);
  char *memPtr = origPtr;
  for (int y = 0; y < tileHeight; y++)
  {
    // Mimic the behaviour of GL_CLAMP_EDGE for the bottom line
    if (y == tile->height)
    {
      unsigned int* memPtrLong = (unsigned int*)memPtr;
      unsigned int* memPtrLong_previous = (unsigned int*)(memPtr - tileWidth * 4);

      for (int x = 0; x < tileWidth; x++)
        memPtrLong[x] = memPtrLong_previous[x] & 0x00FFFFFF;

      continue;
    }

    lastPixelWasTransparent = false;
    const uint8_t *scanline_before = bitmap->GetScanLine(y + tile->y - 1);
    const uint8_t *scanline_at     = bitmap->GetScanLine(y + tile->y);
    const uint8_t *scanline_after  = bitmap->GetScanLine(y + tile->y + 1);
    for (int x = 0; x < tileWidth; x++)
    {

/*    if (target->_colDepth == 15)
      {
        unsigned short* memPtrShort = (unsigned short*)memPtr;
        unsigned short* srcData = (unsigned short*)&bitmap->GetScanLine[y + tile->y][(x + tile->x) * 2];
        if (*srcData == MASK_COLOR_15) 
        {
          if (target->_opaque)  // set to black if opaque
            memPtrShort[x] = 0x8000;
          else if (!usingLinearFiltering)
            memPtrShort[x] = 0;
          // set to transparent, but use the colour from the neighbouring 
          // pixel to stop the linear filter doing black outlines
          else
          {
            unsigned short red = 0, green = 0, blue = 0, divisor = 0;
            if (x > 0)
              get_pixel_if_not_transparent15(&srcData[-1], &red, &green, &blue, &divisor);
            if (x < tile->width - 1)
              get_pixel_if_not_transparent15(&srcData[1], &red, &green, &blue, &divisor);
            if (y > 0)
              get_pixel_if_not_transparent15((unsigned short*)&bitmap->GetScanLine[y + tile->y - 1][(x + tile->x) * 2], &red, &green, &blue, &divisor);
            if (y < tile->height - 1)
              get_pixel_if_not_transparent15((unsigned short*)&bitmap->GetScanLine[y + tile->y + 1][(x + tile->x) * 2], &red, &green, &blue, &divisor);
            if (divisor > 0)
              memPtrShort[x] = ((red / divisor) << 10) | ((green / divisor) << 5) | (blue / divisor);
            else
              memPtrShort[x] = 0;
          }
          lastPixelWasTransparent = true;
        }
        else
        {
          memPtrShort[x] = 0x8000 | (algetr15(*srcData) << 10) | (algetg15(*srcData) << 5) | algetb15(*srcData);
          if (lastPixelWasTransparent)
          {
            // update the colour of the previous tranparent pixel, to
            // stop black outlines when linear filtering
            memPtrShort[x - 1] = memPtrShort[x] & 0x7FFF;
            lastPixelWasTransparent = false;
          }
        }
      }
      else if (target->_colDepth == 32)
*/
      {
        unsigned int* memPtrLong = (unsigned int*)memPtr;

        if (x == tile->width)
        {
          memPtrLong[x] = memPtrLong[x - 1] & 0x00FFFFFF;
          continue;
        }

        unsigned int* srcData = (unsigned int*)&scanline_at[(x + tile->x) << 2];
        if (*srcData == MASK_COLOR_32)
        {
          if (target->_opaque)  // set to black if opaque
            memPtrLong[x] = 0xFF000000;
          else if (!usingLinearFiltering)
            memPtrLong[x] = 0;
          // set to transparent, but use the colour from the neighbouring 
          // pixel to stop the linear filter doing black outlines
          else
          {
            unsigned int red = 0, green = 0, blue = 0, divisor = 0;
            if (x > 0)
              get_pixel_if_not_transparent32(&srcData[-1], &red, &green, &blue, &divisor);
            if (x < tile->width - 1)
              get_pixel_if_not_transparent32(&srcData[1], &red, &green, &blue, &divisor);
            if (y > 0)
              get_pixel_if_not_transparent32((unsigned int*)&scanline_before[(x + tile->x) << 2], &red, &green, &blue, &divisor);
            if (y < tile->height - 1)
              get_pixel_if_not_transparent32((unsigned int*)&scanline_after[(x + tile->x) << 2], &red, &green, &blue, &divisor);
            if (divisor > 0)
              memPtrLong[x] = ((red / divisor) << 16) | ((green / divisor) << 8) | (blue / divisor);
            else
              memPtrLong[x] = 0;
          }
          lastPixelWasTransparent = true;
        }
        else if (hasAlpha)
        {
          memPtrLong[x] = D3DCOLOR_RGBA(algetr32(*srcData), algetg32(*srcData), algetb32(*srcData), algeta32(*srcData));
        }
        else
        {
          memPtrLong[x] = D3DCOLOR_RGBA(algetr32(*srcData), algetg32(*srcData), algetb32(*srcData), 0xff);
          if (lastPixelWasTransparent)
          {
            // update the colour of the previous tranparent pixel, to
            // stop black outlines when linear filtering
            memPtrLong[x - 1] = memPtrLong[x] & 0x00FFFFFF;
            lastPixelWasTransparent = false;
          }
        }
      }
    }

    memPtr += tileWidth * 4;
  }

  glBindTexture(GL_TEXTURE_2D, tile->texture);
  glTexSubImage2D(GL_TEXTURE_2D, 0, 0, 0, tileWidth, tileHeight, GL_RGBA, GL_UNSIGNED_BYTE, origPtr);

  free(origPtr);
}`

**Comment by toojays** [comment]
From #510:
> @toojays could you also look at my findings in #574 regarding vertical texture lines you experienced? I am interested to know if the suggested fix works.
> (Still not 100% sure that what I fixed is the same thing though because I only reproduced the glitch with linear scaling filter but not StdScale one)

I couldn't reproduce it with StdScale in Gemini Rue, but I could reproduce it in Lamplight City. See just to the left of the protagonist in the following:
![lamplight-city-stdscale-artifact](https://user-images.githubusercontent.com/378690/50828335-006f7180-1391-11e9-85af-d932882e717a.png)

With the change you suggested, i.e.
```diff
--- a/Engine/gfx/ali3dogl.cpp
+++ b/Engine/gfx/ali3dogl.cpp
@@ -1694,6 +1694,16 @@ void OGLGraphicsDriver::UpdateTextureRegion(OGLTextureTile *tile, Bitmap *bitmap
   int pitch = tileWidth * sizeof(int);
   BitmapToVideoMem(bitmap, hasAlpha, &fixedTile, target, memPtr, pitch, usingLinearFiltering);
 
+  if (tile->width < tileWidth)
+  {
+      for (int y = 0; y < tileHeight; y++)
+      {
+          unsigned int* memPtrLong = (unsigned int*)(memPtr + y * pitch + tile->width * sizeof(int));
+          unsigned int* memPtrLong_previous = (unsigned int*)(memPtr + y * pitch + (tile->width - 1) * sizeof(int));
+          *memPtrLong = *memPtrLong_previous & 0x00FFFFFF;
+      }
+  }
+
   // Mimic the behaviour of GL_CLAMP_EDGE for the bottom line
   if (tile->height < tileHeight)
   {
```

This artifact goes away. It also fixes the Gemini Rue case I reported in #510.

FWIW I find it clearer if the above defines ``memPtrLong_previous = memPtrLong - 1``.

I did also try redefining ``GL_CLAMP`` to ``GL_CLAMP_TO_EDGE``, but that didn't seem to make a difference. Maybe we need to skip adding the extra pixel to ``tileWidth`` and ``tileHeight`` in that case? I didn't try that.

**Comment by toojays** [comment]
Oh, and if you go ahead and change ``UpdateTextureRegion()``, you might as well remove the unused variable ``lastPixelWasTransparent``.

**Comment by toojays** [comment]
One last post from me tonight. https://stackoverflow.com/questions/19611745/opengl-black-lines-in-between-tiles suggests we need a border all the way around. Why does it work to only have a border on one side of each axis?

**Comment by ivan-mogilko** [comment]
> One last post from me tonight. https://stackoverflow.com/questions/19611745/opengl-black-lines-in-between-tiles suggests we need a border all the way around. Why does it work to only have a border on one side of each axis?

So, as much as I understand this, the stackoverflow question was regarding the **texture atlas**, that is when there are multiple tiles/sprites per texture. In our case we have only 1 sprite per texture and it is "physically" limited from top-left by the texture's surface. The whole problem in our case originates from texture size being larger than the sprite, stretching a little further to right/bottom.

> I did also try redefining GL_CLAMP to GL_CLAMP_TO_EDGE, but that didn't seem to make a difference. Maybe we need to skip adding the extra pixel to tileWidth and tileHeight in that case? I didn't try that.

I tried all variants and GL_CLAMP_TO_EDGE without extra pixels did not work on Windows, but haven't yet tried on Linux. I think it may not work on Windows because of OpenGL interface version we use (but not sure).

**Comment by 256colour** [comment]
@ivan-mogilko I used the code updated in the Mac and Linux a few days ago, and just heard back from our testers - this definitely fixed the lines on their end. :)

I'll patch in @toojays modification to `memPtrLong_previous = memPtrLong - 1` as well.

**Comment by ivan-mogilko** [comment]
> @toojays Could you post your modified code using memPtrLong_previous = memPtrLong - 1? Would like to make the same change in my fork. Thanks!

You simply replace line 
`unsigned int* memPtrLong_previous = (unsigned int*)(memPtr + y * pitch + (tile->width - 1) * sizeof(int));`
with 
`unsigned int* memPtrLong_previous = memPtrLong - 1;`

**Comment by ivan-mogilko** [comment]
Pushed a fix: b3195e4a3e36b9a4b081665e87a8134e3ac01bc5

**Comment by 256colour** [comment]
@ivan-mogilko Great, thanks. Just updated my fork.

**Comment by toojays** [comment]
@ivan-mogilko 
> I tried all variants and GL_CLAMP_TO_EDGE without extra pixels did not work on Windows, but haven't yet tried on Linux. I think it may not work on Windows because of OpenGL interface version we use (but not sure).

My reading suggested that Windows' software OpenGL renderer defaults to OpenGL 1.1 unless you ask for something different (and AGS does not, AFAICS). GL_CLAMP_TO_EDGE wasn't supported until OpenGL 1.2. But on my (Linux) box, the default OpenGL context version seems to be 3.0. It gets logged at AGS startup like:
```
Running OpenGL: 3.0 Mesa 18.2.2
```
So GL_CLAMP_TO_EDGE should be supported here.

I also read that on Windows you may get a different OpenGL context version depending on the drivers you have installed.

This is not at all my area of expertise, thanks for your reply to my question about the border.

**Comment by ivan-mogilko** [comment]
>> One last post from me tonight. https://stackoverflow.com/questions/19611745/opengl-black-lines-in-between-tiles suggests we need a border all the way around. Why does it work to only have a border on one side of each axis?

> So, as much as I understand this, the stackoverflow question was regarding the texture atlas, that is when there are multiple tiles/sprites on one texture. In our case we have only 1 sprite per texture and it is "physically" limited at top-left by the texture's surface. The whole problem in our case originates from texture size being larger than the sprite, stretching a little further to right/bottom, this is why we need to paint right and bottom border.


Testing unrelated issue in a random old game I noticed that some sprites actually have black/white lines from top-left too even though I thought that should not be a case in AGS. So perhaps they DO need a full border. I will investigate more and make commit if that fixes it.

For the reference, the game where I found this is "Apprentice 2". I don't know if it's being a native letterbox game or full resolution having 4:3 aspect ratio has anything to do with this, but maybe... (I haven't seen this effect in newer games since but that could be coincidence)

**Comment by sonneveld** [comment]
> My reading suggested that Windows' software OpenGL renderer defaults to OpenGL 1.1 unless you ask for something different (and AGS does not, AFAICS). GL_CLAMP_TO_EDGE wasn't supported until OpenGL 1.2. 

I know it'll be a pain, but I suggest we stick with OpenGL 2.0 until we can rewrite our code to not use the fixed pipeline. Macos does not support opengl 3 or above unless you use only the Core Profile.

**Comment by sonneveld** [comment]
> @256colour : I'm working on the Mac OpenGL build of the game found in the "osx-nick" branch of Janet's AGS fork. I'm trying to integrate the patch, but I could use some help with modifying the method

Have you tried my sdl branch? The opengl code is similar but it relies on SDL to open windows so it might be a little more reliable.

**Comment by ivan-mogilko** [comment]
>  I suggest we stick with OpenGL 2.0 until we can rewrite our code to not use the fixed pipeline. Macos does not support opengl 3 or above unless you use only the Core Profile.

It looks like Android has similar issue, because I could not make shaders work even after fixing them for GL ES.

**Comment by ivan-mogilko** [comment]
Alright, this seem to fix remaining issue: f2936359958646a20f24316adc5d9c9d87179796

To allow this fix I had to reposition game sprites by 1 texel offset on a texture to make space for the left/top edges (except when sprites fit perfectly on a texture size, or there's only spare 1 row/columns of texels left).
